library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,165 @@
+[JLBP-6] Package and artifact renaming rules
+--------------------------------------------
+
+When a library B in Java depends on another library A through the Maven
+repository system, there are two types of handles that library B needs to use to
+access classes in library A:
+
+1. The Maven coordinates of library A, following the form
+   `${groupId}:${artifactId}:${version}`, for example
+   `com.google.guava:guava:26.0-jre`. The Maven coordinates are used to pull the
+   library's jar from Maven Central. For each pair of group ID and artifact ID
+   (hereafter referenced as "Maven ID"), the user's build system (for example
+   Maven or Gradle) will make a final selection of only one version to put on
+   the classpath. The different build systems use different rules for which
+   version to select.
+2. The fully-qualified class names of the classes in library A. These classes
+   will generally share a Java package (the package as defined in `package`
+   statements, for example `package com.google.common.collect`). The classpath,
+   which is formed from Maven artifacts and possibly non-Maven sources, will be
+   searched for each fully-qualified class name at runtime. In order to simplify
+   the discussion, we will only talk about the Java package, assuming that the
+   classes of the library will be located in that package or subpackages.
+
+When breaking changes are introduced to library A between major version 1 and
+major version 2, a choice needs to be made: to rename or not rename? This
+question applies to both items listed above, the Maven ID (1) and the Java
+package (2).
+
+Recommendations:
+
+- *If* the new library surface is delivered under a new Java package, either use
+  a different Maven ID (different group ID or artifact ID) or bundle the old and
+  new packages together in the under the original Maven ID.
+- *If* the breaking change is made in-place and the Java package is kept the
+  same, use the same Maven ID (same group ID and same artifact ID).
+- There is a tradeoff: if an extremely large library (for example thousands of
+  classes) breaks a single method, the cost of a package rename will be higher
+  than dealing with the fallout of dealing with diamond dependency conflicts for
+  that particular method. The larger the portion of a library that breaks
+  between major versions, the higher the value is from renaming the Java
+  package.
+- Whether making breaking changes or making no breaking changes, don't publish
+  the same classes under multiple Maven IDs. Or put another way: Once you have
```

Comment:
Any idea how often this comes up? If it's a nod at guava_jdk5, it may be a bit oblique. But you may have other examples in mind.

Reply:
Yes, guava-jdk5 is one example, but I want to highlight it because people may not be careful with this if they move code & artifacts across source repos. Basically, you need to keep the same Maven ID even if you're bundling it in a new place. Should I mention that here?

Reply:
Thinking about this more, I realize that I have seen other cases in which people fork or repackage existing projects without changing the package. (People have released per-package Guava artifacts, for example, and various people have made jsr305 available.)

Maybe there's a way to re-frame this to cover that, too? The current phrasing talks in terms of "you" as if you're the only person who did any publishing, but I'm thinking of scenarios in which multiple people publish the same classes.

All that said, we have much bigger fish to fry than the phrasing here, so feel free to just keep what you have. I may have just interpreted it through the lens of guava_jdk5. Other readers are less likely to have that particular lens.

Reply:
That's a very good point: third party republishing with a different group ID and often the same artifact ID, same package, and often ACTUAL INCOMPATIBLE BREAKING CHANGES. I need to think about that one. Don't block on me though. This may well require a separate item.